statistics/getstats.go

```diff
@@ -0,0 +1,171 @@
+)
+
+type dataBuffer struct {
+	data                         []data
+	slidingFactorIndex, old, new int //sliding factor specifies how many data values to include over each sliding window
+}
+
+// data holds the timestamp and the value (actual data)
+type data struct {
+	ts    time.Time
+	value float64
+}
+
+const (
+	count                  = "count"
+	mean                   = "mean"
+	sum                    = "sum"
+	median                 = "median"
+	minimum                = "minimum"
+	maximum                = "maximum"
+	rangeval               = "rangeval"
+	variance               = "variance"
+	standarddeviation      = "standarddeviation"
+	mode                   = "mode"
+	kurtosis               = "kurtosis"
+	skewness               = "skewness"
+	trimean                = "trimean"
+	firstquartile          = "firstquartile"
+	thirdquartile          = "thirdquartile"
+	quartilerange          = "quartilerange"
+	secondpercentile       = "secondpercentile"
+	ninthpercentile        = "ninthpercentile"
+	twentyfifthpercentile  = "twentyfifthpercentile"
+	seventyfifthpercentile = "seventyfifthpercentile"
+	ninetyfirstpercentile  = "ninetyfirstpercentile"
+	ninetyeighthpercentile = "ninetyeighthpercentile"
+	ninetyninthpercentile  = "ninetyninthpercentile"
+	ninetyfifthpercentile  = "ninetyfifthpercentile"
+)
+
+var (
+	statList = []string{count, mean, sum, median, minimum, maximum, rangeval, variance, standarddeviation, mode, kurtosis, skewness, trimean, firstquartile, thirdquartile, quartilerange,
+		secondpercentile, ninthpercentile, twentyfifthpercentile, seventyfifthpercentile, ninetyfirstpercentile, ninetyeighthpercentile, ninetyfifthpercentile, ninetyninthpercentile}
+)
+
+func (b *dataBuffer) Insert(value float64, ts time.Time) {
+	// sort by timestamp before inserting
+	sort.Sort(byTimestamp(b.data))
+	if len(b.data) < cap(b.data) {
+		b.data = append(b.data, data{value: value, ts: ts})
+	} else {
+		// replace older value
+		b.data[0].ts, b.data[0].value = ts, value
+	}
+}
+
+func (d *dataBuffer) GetStats(stats []string, ns []string) ([]plugin.Metric, error) {
+	if len(d.data) == 0 {
+		return nil, nil
+	}
+	var results []plugin.Metric
+
+	// Namespace prefix and tags are common for every stats
+	nsPrefix := append([]string{"intel", "statistics"}, ns...)
+	tags := d.GetTags()
+
+	// statistics are stored in a map
+	statMap := make(result)
+
+	// sort by value is required before calculating statistics
+	sort.Sort(byValue(d.data))
+
+	// generate config option map
+	opts, err := d.SetConfigOption(stats)
+	if err != nil {
+		return nil, err
+	}
+
+	// Calcul statistics
+	for stat, opt := range opts {
+
+		if _, ok := statMap[stat]; !ok {
+			opt(statMap)
+		}
+		newStat := statMap[stat]
+
+		// create the metric we just calculated
+		namespace := plugin.NewNamespace(nsPrefix...).AddStaticElement(stat)
+		switch newStat.(type) {
+		case float64:
+			createMetric(&results, newStat.(float64), tags, namespace)
+		case int:
```

Comment:
This is converting statistics to metric, and we generate only those types of statistics.